compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ConstantValueChecker.java

```diff
@@ -0,0 +1,131 @@
+import org.ballerinalang.util.diagnostic.DiagnosticCode;
+import org.wso2.ballerinalang.compiler.semantics.model.symbols.BSymbol;
+import org.wso2.ballerinalang.compiler.semantics.model.symbols.SymTag;
+import org.wso2.ballerinalang.compiler.tree.BLangIdentifier;
+import org.wso2.ballerinalang.compiler.tree.BLangNodeVisitor;
+import org.wso2.ballerinalang.compiler.tree.expressions.BLangExpression;
+import org.wso2.ballerinalang.compiler.tree.expressions.BLangLiteral;
+import org.wso2.ballerinalang.compiler.tree.expressions.BLangRecordLiteral;
+import org.wso2.ballerinalang.compiler.tree.expressions.BLangRecordLiteral.BLangMapLiteral;
+import org.wso2.ballerinalang.compiler.tree.expressions.BLangRecordLiteral.BLangRecordKeyValue;
+import org.wso2.ballerinalang.compiler.tree.expressions.BLangSimpleVarRef;
+import org.wso2.ballerinalang.compiler.util.CompilerContext;
+import org.wso2.ballerinalang.compiler.util.diagnotic.BLangDiagnosticLog;
+
+/**
+ * @since 0.990.4
+ */
+public class ConstantValueChecker extends BLangNodeVisitor {
+
+    private static final CompilerContext.Key<ConstantValueChecker> CONSTANT_VALUE_RESOLVER_KEY =
+            new CompilerContext.Key<>();
+
+    private BLangDiagnosticLog dlog;
+
+    private BLangExpression referenceExpression;
+    private BLangIdentifier keyIdentifier;
+
+
+    private ConstantValueChecker(CompilerContext context) {
+        context.put(CONSTANT_VALUE_RESOLVER_KEY, this);
+
+        this.dlog = BLangDiagnosticLog.getInstance(context);
+    }
+
+    public static ConstantValueChecker getInstance(CompilerContext context) {
+        ConstantValueChecker constantValueResolver = context.get(CONSTANT_VALUE_RESOLVER_KEY);
+        if (constantValueResolver == null) {
+            constantValueResolver = new ConstantValueChecker(context);
+        }
+        return constantValueResolver;
+    }
+
+    void checkValue(BLangExpression referenceExpression, BLangIdentifier keyIdentifier, BLangRecordLiteral mapLiteral) {
+        this.referenceExpression = referenceExpression;
+        this.keyIdentifier = keyIdentifier;
+
+        mapLiteral.accept(this);
+    }
+
+    @Override
+    public void visit(BLangRecordLiteral recordLiteral) {
+        // Iterate through all key-value pairs in the record literal.
+        for (BLangRecordKeyValue keyValuePair : recordLiteral.keyValuePairs) {
+            //  Get the key.
+            Object key = ((BLangLiteral) keyValuePair.key.expr).value;
+
+            // If the key is equal to the value of the keyIdentifier, that means the key which we are looking for is
+            // in the record literal.
+            if (!key.equals(keyIdentifier.value)) {
+                continue;
+            }
+
+            // Since we are looking for a literal which can be used as at compile time, it should be a literal.
+            NodeKind nodeKind = keyValuePair.valueExpr.getKind();
+            if (nodeKind == NodeKind.LITERAL || nodeKind == NodeKind.NUMERIC_LITERAL ||
```

Comment:
Check the possibility of using the visitor pattern.

Reply:
I checked whether visitor pattern can be used here, but it is not feasible and does not offer any advantage over the current implementation.